File: decorator-pattern/java/src/RedShapeDecorator.java
package src;

// 具体装饰器1
public class RedShapeDecorator extends ShapeDecorator {

  public RedShapeDecorator(Shape decoratedShape) {
    super(decoratedShape);
  }

  @Override
  public void draw() {
    decoratedShape.draw();
    setRedColor(decoratedShape);
  }

  private void setRedColor(Shape decoratedShape) {
    System.out.println(
      "RedShapeDecorator::setRedColor() " + decoratedShape.getClass().getName()
    );
  }
}

File: decorator-pattern/go/src/RedShapeDecorator.go
package src

import "fmt"

// 具体装饰器1，实现基础装饰器接口
type RedShapeDecorator struct {
  DecoratedShape Shape
}

func (r *RedShapeDecorator) Draw() {
  r.DecoratedShape.Draw()
  r.SetRedColor(r.DecoratedShape)
}

func (r *RedShapeDecorator) SetRedColor(decoratedShape Shape) {
  fmt.Println("RedShapeDecorator::setRedColor() " + decoratedShape.GetName())
}
